docs/arch/adr-2504-limit-concurrent-merge-resources.md

```diff
@@ -0,0 +1,31 @@
+# ADR 2311: Limit Concurrent Calls to MergeResources Stored Procedure
+Labels: [SQL](https://github.com/microsoft/fhir-server/labels/Area-SQL)
+
+## Context
+The MergeResources stored procedure is critical for writing FHIR resources to store efficiently. However, excessive concurrent calls to this procedure can degrade database performance and significantly increase write latencies. To mitigate this, we need a mechanism to limit concurrent executions of MergeResources.
+
+## Decision
+We will implement infrastructure to set the application name in the connection string specifically to "MergeResources". Additionally, we will modify the MergeResourceBeginTransaction stored procedure to include a concurrency check using the following logic:
+
+```sql
+SELECT count(*) FROM sys.dm_exec_sessions WHERE status <> 'sleeping' AND program_name = 'MergeResources'
+```
+
+If the count exceeds the configured threshold defined by the parameter `MergeResources.OptimalConcurrentCalls`, the stored procedure will raise an exception. This exception will be caught by the FHIR server code, triggering a retry mechanism with exponential backoff.
+
+Initially, this change will only introduce the infrastructure and exception handling. A subsequent work item will implement logic to return HTTP 429 (Too Many Requests) responses to API callers when concurrency limits are exceeded.
+
+## Status
+Accepted
```

Comment:
Can you add a link to Sergey's PR? I think it is helpful to connect all the relevant parts so that it is easier to track down everything if needed.

Reply:
Three Commits:

https://github.com/microsoft/fhir-server/commit/132cd07f97d47ee82dc5c5bd7799b41a990c1edf

https://github.com/microsoft/fhir-server/commit/a983d46085b2c1c8cc82195ce97516136a59b102

https://github.com/microsoft/fhir-server/commit/71d34ac906725fa84ff66ccb3d7086cec53a87fa